ui/src/criteria/attribute.tsx

```diff
@@ -81,23 +141,149 @@ function AttributeEdit(props: AttributeEditProps) {
+  if (!isValid(integerBoundsHint) && enumHintValues?.length === 0) {
     return (
       <Typography>
         No information for attribute {props.data.attribute}.
       </Typography>
     );
   }
+
+  if (
+    integerBoundsHint?.max !== null &&
+    integerBoundsHint?.max !== undefined &&
+    integerBoundsHint?.min !== null &&
+    integerBoundsHint?.min !== undefined
+  ) {
+    const minBound = integerBoundsHint?.min;
+    const maxBound = integerBoundsHint?.max;
+
+    const [minInputValue, setMinInputValue] = React.useState<number>(minBound);
+    const [maxInputValue, setMaxInputValue] = React.useState<number>(maxBound);
+    const [minValue, setMinValue] = React.useState<number>(minBound);
+    const [maxValue, setMaxValue] = React.useState<number>(maxBound);
+
+    const updateValue = (newMin: number, newMax: number) => {
+      setMinValue(newMin);
+      setMaxValue(newMax);
+      props.dispatchFn(
+        produce(props.data, (data) => {
+          data.min = newMin;
+          data.max = newMax;
+        })
+      );
+    };
+
+    const handleChange = (event: Event, newValue: number | number[]) => {
+      newValue = newValue as number[];
+      const [newMin, newMax] = newValue;
+      setMinInputValue(newMin);
+      setMaxInputValue(newMax);
+      updateValue(newMin, newMax);
+    };
+
+    const handleLeftInputChange = (
+      event: React.ChangeEvent<HTMLInputElement>
+    ) => {
+      let newMin =
+        event.target.value === "" ? minBound : Number(event.target.value);
+      setMinInputValue(newMin);
+
+      newMin = newMin < minBound ? minBound : newMin;
+      newMin = newMin > maxValue ? maxValue : newMin;
+      updateValue(newMin, maxValue);
+    };
+    const handleLeftInputBlur = () => {
+      let newMin = minInputValue;
+      newMin = newMin < minBound ? minBound : newMin;
+      newMin = newMin > maxValue ? maxValue : newMin;
+      setMinInputValue(newMin);
+    };
+
+    const handleRightInputChange = (
+      event: React.ChangeEvent<HTMLInputElement>
+    ) => {
+      let newMax =
+        event.target.value === "" ? maxBound : Number(event.target.value);
+      setMaxInputValue(newMax);
+
+      newMax = newMax < minValue ? minValue : newMax;
+      newMax = newMax > maxBound ? maxBound : newMax;
+      updateValue(minValue, newMax);
+    };
+    const handleRightInputBlur = () => {
+      let newMax = maxInputValue;
+      newMax = newMax < minValue ? minValue : newMax;
+      newMax = newMax > maxBound ? maxBound : newMax;
+      setMaxInputValue(newMax);
+    };
+
+    return (
+      <Box sx={{ width: "30%", minWidth: 100 }}>
+        <Grid container spacing={2} alignItems="center">
+          <Grid item>
+            <Input
+              value={minInputValue}
+              size="small"
+              onChange={handleLeftInputChange}
+              onBlur={handleLeftInputBlur}
+              inputProps={{
+                step: ((maxBound - minBound) / 20) | 0,
```

Comment:
Did you mean to put `| 0` here, i.e. bitwise OR with 0? Is this to force an integer?  Math.floor() is probably more readable for this purpose

Reply:
Yes. I searched online to find a way transfer it into integer. They give me this. Nice to learn the math.floor

Reply:
You should use `Math.floor` here instead of `| 0` like CH suggested.